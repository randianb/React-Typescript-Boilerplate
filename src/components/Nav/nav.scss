.navbar {
  background: #ccc;
  padding: 1rem;
  text-align: right;
  a {
    display: inline-block;
    background: #efef;
    text-decoration: none;
    color: #333;
    margin: 0 0.5rem;
    padding: 0.5rem;
    &:hover {
      color: #ccc;
    }
  }
}
