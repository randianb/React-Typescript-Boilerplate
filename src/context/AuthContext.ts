import React from 'react';

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
