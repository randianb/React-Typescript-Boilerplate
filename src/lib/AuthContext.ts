import React from 'react';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export interface AuthContextInterface {
  checkingSession: boolean;
  token: string | null;
  idToken: string | null;
  expiresAt: number | null;
  isAuthenticated: boolean;
  handleAuthentication: () => void;
  handleLogin: () => void;
  handleLogout: () => void;
}

export const authContextDefaults: AuthContextInterface = {
  checkingSession: false,
  expiresAt: null,
  token: null,
  idToken: null,
  isAuthenticated: false,
  handleAuthentication: () => null,
  handleLogin: () => null,
  handleLogout: () => null
};

export const AuthContext = React.createContext<AuthContextInterface>(
  authContextDefaults
);
