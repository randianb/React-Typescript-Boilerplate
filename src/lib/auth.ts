import auth0 from 'auth0-js';

export interface AuthProfile {
  at_hash?: string;
  aud?: string;
  exp?: number;
  iat?: number;
  iss?: string;
  nonce?: string;
  sub?: string;
  locale?: string;
  name?: string;
  picture?: string;
  email?: string;
}

export class Auth {
  private accessToken: string | null = null;
  private idToken: string | null = null;
  private expiresAt: number = 0;
  private profile: AuthProfile | null = null;
  private auth0 = new auth0.WebAuth({
    domain: process.env.REACT_APP_AUTH_DOMAIN || '',
    clientID: process.env.REACT_APP_AUTH_CLIENT_ID || '',
    redirectUri:
      process.env.REACT_APP_AUTH_CALLBACK || 'http://localhost:3000/callback',
    responseType: 'token id_token',
    scope: 'openid profile email'
  });

  constructor() {
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
    this.handleAuthentication = this.handleAuthentication.bind(this);
    this.isAuthenticated = this.isAuthenticated.bind(this);
    this.getAccessToken = this.getAccessToken.bind(this);
    this.getIdToken = this.getIdToken.bind(this);
    this.silentAuth = this.silentAuth.bind(this);
    this.getProfile = this.getProfile.bind(this);
  }

  public login() {
    this.auth0.authorize();
  }

  public handleAuthentication(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.auth0.parseHash((err, authResult) => {
        if (err) {
          return reject(err);
        }
        if (!authResult || !authResult.idToken) {
          return reject(err);
        }
        this.setSession(authResult);
        resolve();
      });
    });
  }

  public getAccessToken() {
    return this.accessToken;
  }

  public getProfile() {
    return this.profile;
  }

  public getIdToken() {
    return this.idToken;
  }

  public silentAuth(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.auth0.checkSession({}, (err, authResult) => {
        if (err) {
          return reject(err);
        }
        this.setSession(authResult);
        resolve();
      });
    });
  }

  public logout() {
    this.accessToken = null;
    this.idToken = null;
    this.expiresAt = 0;
    this.profile = null;

    this.auth0.logout({
      returnTo: process.env.REACT_APP_AUTH_RETURN_URL,
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID
    });
  }

  public isAuthenticated() {
    return new Date().getTime() < this.expiresAt;
  }

  private async setSession(authResult: any) {
    const expiresAt = authResult.expiresIn * 1000 + new Date().getTime();
    this.accessToken = authResult.accessToken;
    this.idToken = authResult.idToken;
    this.profile = authResult.idTokenPayload;
    this.expiresAt = expiresAt;
  }
}
