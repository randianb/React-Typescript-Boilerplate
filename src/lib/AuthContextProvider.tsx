import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';
import { withRouter, RouteComponentProps } from 'react-router';
import auth0 from 'auth0-js';
class AuthContextProvider extends React.Component<RouteComponentProps> {
  private auth0: auth0.WebAuth;

  public state: AuthContextInterface = {
    token: null,
    idToken: null,
    expiresAt: 0,
    checkingSession: true,
    isAuthenticated: false,
    handleAuthentication: this.handleAuthentication.bind(this),
    handleLogin: this.handleLogin.bind(this),
    handleLogout: this.handleLogout.bind(this)
  };

  constructor(props: RouteComponentProps) {
    super(props);
    this.auth0 = new auth0.WebAuth({
      domain: process.env.REACT_APP_AUTH_DOMAIN || '',
      clientID: process.env.REACT_APP_AUTH_CLIENT_ID || '',
      redirectUri: process.env.REACT_APP_AUTH_CALLBACK,
      responseType: 'token id_token',
      scope: 'openid'
    });
  }

  public handleLogin() {
    this.auth0.authorize();
  }

  public handleAuthentication() {
    this.auth0.parseHash((err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        this.setSession(authResult);
      } else if (err) {
        console.log('handleAuthentication error', err);
      }
      this.props.history.push('/');
    });
  }

  public getToken() {
    return this.state.token;
  }

  public getIdToken() {
    return this.state.idToken;
  }

  public setSession(authResult: auth0.Auth0DecodedHash) {
    localStorage.setItem('isLoggedIn', 'true');
    const expiresAt = authResult.expiresIn
      ? authResult.expiresIn * 1000 + new Date().getTime()
      : 0;
    this.setState({
      token: authResult.accessToken || null,
      idToken: authResult.idToken || null,
      expiresAt,
      isAuthenticated: true,
      checkingSession: false
    });
  }

  public renewSession() {
    this.auth0.checkSession({}, (err, authResult) => {
      if (authResult && authResult.accessToken && authResult.idToken) {
        this.setSession(authResult);
      } else if (err) {
        console.log('renewSession error', err);
      }
      this.setState({
        checkingSession: false
      });
    });
  }

  public handleLogout() {
    this.setState({
      token: null,
      idToken: null,
      expiresAt: 0,
      isAuthenticated: false
    });

    localStorage.removeItem('isLoggedIn');

    this.auth0.logout({
      returnTo: window.location.origin
    });

    this.props.history.push('/');
  }

  public isAuthenticated() {
    const expiresAt = this.state.expiresAt || 0;
    return new Date().getTime() < expiresAt;
  }

  public componentDidMount() {
    this.renewSession();
  }

  public render() {
    return (
      <AuthContext.Provider value={this.state}>
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export const AppContextProviderWithRoutes = withRouter(AuthContextProvider);
