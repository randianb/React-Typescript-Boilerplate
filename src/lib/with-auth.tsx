import React from 'react';
import { AuthContext, AuthContextInterface } from './AuthContext';

export interface AuthProps {
  auth: AuthContextInterface;
}

export function withAuth<P extends AuthProps>(Component: React.ComponentType<P>) {
  return function AuthedComponent(props: Pick<P, Exclude<keyof P, keyof AuthProps>>) {
    return (
      <AuthContext.Consumer>
        {auth => <Component {...props as P} auth={auth} />}
      </AuthContext.Consumer>
    );
  };
}
