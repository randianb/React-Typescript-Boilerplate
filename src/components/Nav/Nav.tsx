import React, { useContext } from 'react';
import { Link } from 'react-router-dom';
import { AuthContext } from '../../lib/AuthContext';
import './nav.scss';

export const Nav = () => {
  const context = useContext(AuthContext);
  return (
    <div className="navbar">
      <Link to="/">Home</Link>
      <Link to="/admin">Admin</Link>
      {context.isAuthenticated && <Link to="/logout">Logout</Link>}
      {!context.isAuthenticated && <Link to="/login">Login</Link>}
    </div>
  );
};
