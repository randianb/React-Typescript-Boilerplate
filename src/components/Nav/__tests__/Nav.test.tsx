import React from 'react';
import { Nav } from '../Nav';
import { render, cleanup } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { AuthContext, authContextDefaults } from '../../../lib/AuthContext';

afterEach(cleanup);

const auth = authContextDefaults;
const TestComponent = (props: any) => {
  return (
    <AuthContext.Provider value={props.auth}>
      <MemoryRouter initialEntries={['/foo']} initialIndex={0}>
        <Nav />
      </MemoryRouter>
    </AuthContext.Provider>
  );
};

describe('Nav', () => {
  it('Renders Nav', () => {
    const { container } = render(<TestComponent auth={auth} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Logged In State', () => {
    auth.isAuthenticated = true;
    auth.profile = {
      name: 'Bob'
    };
    const { container } = render(<TestComponent auth={auth} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Logged out State', () => {
    auth.isAuthenticated = false;
    const { container } = render(<TestComponent auth={auth} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
