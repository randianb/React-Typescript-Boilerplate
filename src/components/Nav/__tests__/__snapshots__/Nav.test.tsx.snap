// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav Logged In State 1`] = `
<div
  class="Navbar"
>
  <span
    id="loggedInUser"
  >
    Logged in as: 
    Bob
  </span>
  <a
    aria-current="page"
    class="active"
    href="/"
  >
    Home
  </a>
  <a
    href="/admin"
  >
    Admin
  </a>
  <a
    href="/logout"
    id="logout-link"
  >
    Logout
  </a>
</div>
`;

exports[`Nav Logged out State 1`] = `
<div
  class="Navbar"
>
  <a
    aria-current="page"
    class="active"
    href="/"
  >
    Home
  </a>
  <a
    href="/admin"
  >
    Admin
  </a>
  <a
    href="/login"
    id="login-link"
  >
    Login
  </a>
</div>
`;

exports[`Nav Renders Nav 1`] = `
<div
  class="Navbar"
>
  <a
    aria-current="page"
    class="active"
    href="/"
  >
    Home
  </a>
  <a
    href="/admin"
  >
    Admin
  </a>
  <a
    href="/login"
    id="login-link"
  >
    Login
  </a>
</div>
`;
